### ISSUE TITLE
Improving flake formatting experience

### ISSUE THREAD

**Issue by GaetanLepage** [NixOS/nixfmt#273]
In order to have your flake formatted with `nixfmt-rfc-style`, you have two options:

1. Setting `formatter = pkgs.nixfmt-rfc-style;`. This solution is concise (ignoring the fact that the name "nixfmt-rfc-style" is not really intuitive compared to simply "nixfmt").
    However, this is not expected to be supported for long as: `Passing directories or non-Nix files (such as ".") is deprecated and will be unsupported soon, please use https://treefmt.com/ instead, e.g. via https://github.com/numtide/treefmt-nix`.
2. Using [treefmt-nix](https://github.com/numtide/treefmt-nix). While it is for sure a great, very flexible and powerful tool, it is quite overkilled if you simply want to format a small nix-only code base. It's major con is having to add both several lines of code and an extra dependency, simply to achieve what the first solution is doing.

I think it should be great to support the existing `formatter = pkgs.nixfmt[-rfc-style];` syntax as it gives a very efficient and user-friendly way to format a flake while keeping it lean code and input wise.

In the long term, I could see the benefit of further "officializing" the `nixfmt` formatter and having a `format = true;` flag to automatically format the flake. This would have to be thought more.

cc @infinisil

**Comment by MattSturgeon** [comment]
That one is kinda awkward, since it was already decided that recursively finding nix files to be formatted is a non-trivial issue that is kinda out-of-scope for a tool that just wants to be able to format _a_ nix file.

IIRC, the main complexities come from ignoring `.git` and respecting `.gitignore`.

Maybe this could be worked around by renaming `pkgs.nixfmt-rfc-style` something like `pkgs.nixfmt-unwrapped` and having a wrapper package that handles the logic required to format a directory recursively?

The wrapper package could of course use an existing lib/tool to implement that. And treefmt users should probably be able to use either the wrapped or unwrapped package.

**Comment by eclairevoyant** [comment]
It seems quite trivial, actually; this decision comes off more like it's just intentionally making it harder to use flakes. Using a random third-party tool just to be able to use the _official_ formatter with _official_ nix is a nonstarter IMO. And there's some irony here that [the nix docs still give the example of setting `formatter.<system>` to `nixfmt`](https://nix.dev/manual/nix/2.24/command-ref/new-cli/nix3-fmt.html#examples)... plus if we're talking about scope, `nix` even having a `fmt` subcommand is a strange scope creep, though it's certainly offtopic here to go further into that rabbithole.

In any case, as amusing as whining about nix* community dysfunction is, since I'd rather just get things done, `fd` is my tool of choice for this sort of thing:

```
fd -t f -e nix -x nixfmt '{}'
```

I specifically chose `fd` over `find`, as `fd` ignores hidden files/dirs and respects `.gitignore` the way one would hope `nix fmt` to operate. Also note this only includes regular files. And I do expect `fd` to be better-maintained, better-documented, and stabler in behaviour and API than `treefmt` or other third-party nix ecosystem tools.

To be able to effectively use `nix fmt` rather than hardcoding the formatter in the command, of course, one can substitute `nixfmt` with `nix fmt` in the command above; alternatively, since I'd rather the search-and-format details be properly encapsulated within the formatter package, I prefer to use a wrapper script around `nixfmt` via `writeShellApplication`:

```nix
  formatter.x86_64-linux =
    let
      pkgs = nixpkgs.legacyPackages.x86_64-linux;
    in
    pkgs.writeShellApplication {
      name = "nixfmt-wrapper";

      runtimeInputs = [
        pkgs.fd
        pkgs.nixfmt-rfc-style
      ];

      text = ''
        fd "$@" -t f -e nix -x nixfmt '{}'
      '';
    };
```

The args that are passed (to the formatter package's executable) appear to be the path(s) that are (im/ex)plicitly passed to `nix fmt`; if no args are explicitly passed, the CWD is implicitly passed. (Arguably it should be the flake enclosing the CWD, since that's the implicit flake that is passed to any other nix3 command by default, but...)

**Comment by eclairevoyant** [comment]
Sure, I'm not entirely disagreeing that having nixfmt behave differently from the other formatters is surprising; I just wanted to provide a straightforward workaround that works today and will continue to work even if directory support is removed from nixfmt entirely.

**Comment by MattSturgeon** [comment]
I think I agree with @eclairevoyant that we should avoid tightly coupling nixfmt with treefmt. I prefer their proposal of a wrapper that uses something more generic like `fd`.

I'd propose having:

- `pkgs.nixfmt`: a wrapper that calls nixfmt using `fd`
- `pkgs.nixfmt-unwrapped`: the current `nixfmt-rfc-style` package.

The wrapper script could either be maintained in nixpkgs, or here in nixfmt. I don't think that's too important.

If the wrapper script is maintained here, then perhaps there's no need for separate wrapper packages?

I also think we should avoid going down the rabbit hole of changing how `nix fmt` behaves, at least for now, because that could open up a tonne of bike shedding...

**Comment by infinisil** [comment]
The ultimate test is whether it works for Nixpkgs development, which the simple `fd` solution wouldn't because it takes too long (iirc multiple minutes). Treefmt is super valuable for this because it avoids running the formatter on files that weren't changed, runs the formatter in parallel, and also allows multiplexing to different formatters.

**Comment by jfly** [comment]
+1 to not having `nixfmt` have the directory recursion logic. Giving people a convenient wrapper makes sense, and if we do it right, the implementation (`fd` vs `treefmt`) would just be a detail that users don't have to think about.

One minor thing to keep in mind: treefmt/treefmt-nix make an effort to have `nix fmt` do sane things:

- `nix fmt`: format the whole flake, *regardless of the current directory* (this requires being able to discover the root of the flake)
- `nix fmt ./path/to/file`: format `./path/to/file` *relative to the current directory*

I believe neither [@eclairevoyant's `fd` solution above](https://github.com/NixOS/nixfmt/issues/273#issuecomment-2563424097) nor [@infinisil's POC `treefmt` wrapper](https://github.com/NixOS/nixfmt/issues/273#issuecomment-2576238653) handle this.

**Comment by infinisil** [comment]
All credit to the POC goes to @dasJ! Updated the meeting notes to point that out :smile:

**Comment by MattSturgeon** [comment]
> format the whole flake, regardless of the current directory (this requires being able to discover the root of the flake)

Correct me if I'm wrong, but with the current nix3 and flake implementation doesn't this require some pretty nasty hacks and/or assumptions?

No matter what treefmt-nix _attempts_, there isn't a deterministic way to accurately & reliably find the flake root. Right?

I assume using treefmt without treefmt-nix doesn't bring this baggage, though?

**Comment by jfly** [comment]
> Correct me if I'm wrong, but with the current nix3 and flake implementation doesn't this require some pretty nasty hacks and/or assumptions?

Yeah, that's what treefmt-nix's `projectRootFile` setting is all about: so it can discover and then [tell `treefmt` where the tree-root is](https://github.com/numtide/treefmt-nix/blob/13c913f5deb3a5c08bb810efd89dc8cb24dd968b/module-options.nix#L196). Sure would be nice if `nix` did this for us.

And now I realize I've steered us towards that rabbit hole of bike shedding you mentioned above...

**Comment by infinisil** [comment]
I've just opened https://github.com/NixOS/nixpkgs/pull/384857 to mostly address this

**Comment by infinisil** [comment]
Let's keep this open for now. Small TODOs:
- Update Nix's `nix fmt` docs: https://github.com/NixOS/nix/blob/master/src/nix/fmt.md
- https://github.com/NixOS/nixfmt/pull/287

**Comment by jfly** [comment]
I am working on this in https://github.com/NixOS/nix.dev/pull/1129

**Comment by jfly** [comment]
Discussed at today's team meeting: we're happy with the technical state of things, but the documentation is not where we want it.

We're tracking that effort in this PR to nix: https://github.com/NixOS/nix/pull/13014.

I'm closing this issue. Feel free to file new issues with future ideas!
